src/litdata/streaming/downloader.py

```diff
@@ -169,8 +169,15 @@ def download_file(self, remote_filepath: str, local_filepath: str) -> None:
         if not os.path.exists(remote_filepath):
             raise FileNotFoundError(f"The provided remote_path doesn't exist: {remote_filepath}")
 
-        if remote_filepath != local_filepath and not os.path.exists(local_filepath):
-            shutil.copy(remote_filepath, local_filepath)
+        try:
+            with FileLock(local_filepath + ".lock", timeout=3 if remote_filepath.endswith(_INDEX_FILENAME) else 0):
+                if remote_filepath != local_filepath and not os.path.exists(local_filepath):
+                    # make an atomic operation to be safe
+                    temp_file_path = local_filepath + ".tmp"
+                    shutil.copy(remote_filepath, temp_file_path)
+                    os.rename(temp_file_path, local_filepath)
```

Comment:
Hi @tchaton,

I realise the merge request is already closed, but I wanted to get back to you on this. I initially also wondered while the lock files are not deleted by the `FileLock` context manager. I looked at their source code to make sure that I did not miss anything and found references to an [Issue](https://github.com/tox-dev/py-filelock/issues/31) and an [SO thread](https://stackoverflow.com/questions/17708885/flock-removing-locked-file-without-race-condition): [here](https://github.com/tox-dev/filelock/blob/85760d7531cfaa6fe9382d950aadf2ebd21364a0/src/filelock/_unix.py#L55).
Deleting the file can introduce a new race condition, as outlined in the SO thread. Looking at the change that originally introduced the RC and got reverted afterwards, it looks similar to what you have introduced now. 

I am uncertain whether this is a huge risk, but I wanted to make you aware in case any other issues are opened that could be caused by this.

Reply:
Let it run and see if you can trigger the same error as before.